### PR TITLE
service discovery: add HTTP endpoints and sdk wrapper

### PR DESCRIPTION
**Pull request by jrasell**
This work continues to build out the service discovery feature by adding the HTTP and Go SDK covering all the publicly available RPC endpoints. This includes a new endpoint as well as updates to the alloc and job endpoints to provide easy lookup of registrations associated to these objects.

The PR is split into commits which group the additions into logical chunks and hope to make it easier to review.

closes https://github.com/hashicorp/team-nomad/issues/262

### CODE REVIEW

File: api/service_registrations.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+// registrations; particularly when used with blocking queries.
+type ServiceRegistrationListStub struct {
+
+	// Services is an array of service registration stubs, grouped by their
+	// namespace. The length should always be one, unless the query is made
+	// using the namespace wildcard parameter.
+	Services []*ServiceRegistrationStub
+
+	CreateIndex uint64
+	ModifyIndex uint64
```
Comment (tgross) [inline api/service_registrations.go:76]: Aren't these values going to be set in the RPC handler on the `QueryMeta` already? It looks like that's being handled in the HTTP endpoint too.

(And if we didn't have these fields, we could avoid having a separate struct for ListStub instead of sending back a slice of stubs, I think?)

**Review by tgross — APPROVED** [review]
LGTM. I did have one question about the shape of the response we're returning from List but I could just be missing something.

File: api/service_registrations.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+)
+
+// ServiceRegistrations is used to query the service endpoints.
+type ServiceRegistrations struct {
+	client *Client
+}
+
+// ServiceRegistration is the fully hydrated object which contains all details
+// of an individual service registration.
+type ServiceRegistration struct {
```
Comment (schmichael) [inline api/service_registrations.go:15]: Might want to explain the difference between this and `api/services.go#Service`. Something like:

```go
// ServiceRegistrations represent individual allocations registered with a Nomad Service.
// The Services struct represents the service stanza in jobspecs and may become ServiceRegistrations
// in Consul or Nomad depending on provider.
```

...that might be a little too wordy. Just want to try to help people get to what they're looking for as quickly as possible since the names alone don't make it obvious.

**Review by schmichael — APPROVED** [review]

File: api/service_registrations.go (hunk 1 above)
Comment (jrasell) [reply api/service_registrations.go:76]: I think you're completely right, thanks! I'll make this change now.
